src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/BinaryMessageParser.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Binary;
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    internal class BinaryMessageParser
+    {
+        private ParserState _state;
+
+        public void Reset()
+        {
+            _state = default(ParserState);
+        }
+
+        public bool TryParseMessage(ref BytesReader buffer, out Message message)
+        {
+            if (_state.Length == null)
+            {
+                var length = buffer.TryReadBytes(sizeof(long))?.ToSingleSpan();
```

Comment:
A lack of `ref this` extension methods makes me hesitant to write too many helpers for this kind of thing unfortunately.